crates/net/discv4/src/proto.rs

```diff
@@ -191,6 +194,72 @@ pub struct Packet {
     pub hash: B256,
 }
 
+/// Represents the `from` field in the `Ping` packet
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, RlpEncodableWrapper)]
+pub struct PingNodeEndpoint(pub NodeEndpoint);
+
+impl alloy_rlp::Decodable for PingNodeEndpoint {
+    #[inline]
+    fn decode(b: &mut &[u8]) -> alloy_rlp::Result<Self> {
+        let alloy_rlp::Header { list, payload_length } = alloy_rlp::Header::decode(b)?;
+        if !list {
+            return Err(alloy_rlp::Error::UnexpectedString);
+        }
+        let started_len = b.len();
+        if started_len < payload_length {
+            return Err(alloy_rlp::Error::InputTooShort);
+        }
+
+        // Geth allows the ipaddr to be possibly empty:
+        // <https://github.com/ethereum/go-ethereum/blob/380688c636a654becc8f114438c2a5d93d2db032/p2p/discover/v4_udp.go#L206-L209>
+        // <https://github.com/ethereum/go-ethereum/blob/380688c636a654becc8f114438c2a5d93d2db032/p2p/enode/node.go#L189-L189>
+        //
+        // Therefore, if we see an empty list instead of a properly formed `IpAddr`, we will
+        // instead use `IpV4Addr::UNSPECIFIED`
+        let address =
+            if *b.first().ok_or(alloy_rlp::Error::InputTooShort)? == alloy_rlp::EMPTY_STRING_CODE {
+                let addr = IpAddr::V4(Ipv4Addr::UNSPECIFIED);
+                b.advance(1);
+                addr
+            } else {
+                alloy_rlp::Decodable::decode(b)?
+            };
+
+        let this = NodeEndpoint {
+            address,
+            udp_port: alloy_rlp::Decodable::decode(b)?,
+            tcp_port: alloy_rlp::Decodable::decode(b)?,
+        };
+        let consumed = started_len - b.len();
+        if consumed != payload_length {
+            return Err(alloy_rlp::Error::ListLengthMismatch {
+                expected: payload_length,
+                got: consumed,
+            });
+        }
+        Ok(Self(this))
+    }
+}
+
+impl From<NodeRecord> for PingNodeEndpoint {
+    fn from(NodeRecord { address, tcp_port, udp_port, .. }: NodeRecord) -> Self {
+        Self(NodeEndpoint { address, tcp_port, udp_port })
+    }
+}
+
+impl From<NodeEndpoint> for PingNodeEndpoint {
+    fn from(value: NodeEndpoint) -> Self {
+        Self(value)
+    }
+}
+
+impl PingNodeEndpoint {
+    /// Creates a new [`PingNodeEndpoint`] from a given UDP address and TCP port.
+    pub const fn from_udp_address(udp_address: &std::net::SocketAddr, tcp_port: u16) -> Self {
+        Self(NodeEndpoint { address: udp_address.ip(), udp_port: udp_address.port(), tcp_port })
+    }
+}
```

Comment:
do we need these?

we only need the `.0`?

Reply:
don't need these